bip-cc.md

```diff
@@ -0,0 +1,207 @@
+increased block space would - all others things equal - bring fee rates down for users.
+
+Specially crafted blocks may be expensive to process, with validation times ranging from a few
+minutes up to more than an hour on lower-end devices. Long block validation times are a nuisance to
+users, increasing the cost to independently fully validate the consensus rules. In addition they can
+be used by miners to attack their competition, creating perverse incentives, centralization
+pressures and leading to reduced network security.
+
+In computing a block's Merkle root, a 64-byte transaction can be interpreted as an intermediate
+node in the tree in addition to a leaf. This makes it possible to fake inclusion proofs by
+pretending a 64-byte block transaction is an inner node, as well as to pretend the inner nodes on
+one level of the tree are the actual block transactions.
+
+Since [bip-0034][BIP34] activation, explicit [bip-0030][BIP30] validation is not necessary until
+block height 1,983,702[^0].  Mandating new coinbase transactions be different from the early
+[bip-0034][BIP34] violations makes it possible to get rid of [bip-0030][BIP30] validation forever.
+Besides its unnecessary cost, another downside of [bip-0030][BIP30] validation is that it cannot be
+performed by Utreexo clients. Finally, leveraging the coinbase transaction's `nLockTime` field
+allows applications to recover the block height corresponding to a coinbase transaction without
+having to parse Script.
+
+## Specification
+
+For all blocks after activation the following new rules apply.
+
+Given a block at height `N`:
+- if `N % 2016` is equal to 0 the timestamp of the block must be set to a value higher than or
+  equal to the value of the timestamp of block at height `N-1` minus 7200;
+- if `N % 2016` is equal to 2015 the timestamp of the block must be set to a value higher than
+  or equal to the value of the timestamp of block at height `N-2015`.
+
+A limit is set on the number of potentially executed signature operations in validating a
+transaction. It applies to all transactions in the block except the coinbase transaction[^1]. For
+each input in the transaction, count the number of `CHECKSIG` and `CHECKMULTISIG` in the input
+scriptSig and previous output's scriptPubKey, including the P2SH redeemScript. The accounting is the
+same as for [bip-0016][BIP16 specs]: a `CHECKSIG`/`CHECKSIGVERIFY` operation accounts for 1 and a
+`CHECKMULTISIG`/`CHECKMULTISIGVERIFY` accounts for the number of public keys associated, or 20 if
+the number of public keys is greater than 16. A `CHECKMULTISIG`/`CHECKMULTISIGVERIFY` not directly
+preceded by a minimally-pushed number between 1 and 16 (included) accounts for 20. If the
+total is strictly higher than 2500, the transaction is invalid.
+
+Transactions whose witness-stripped serialized size is exactly 64 bytes are made invalid.
+
+The coinbase transaction's `nLockTime` field must be set to the height of the block minus 1[^2]
```

Comment:
nLockTime is the most ideal location to put an extranonce, so it seems like a bad idea to burn it for no good reason.

Reply:
It would have been better to bring up conceptual feedback on the mailing list. "For no good reason" is unnecessarily aggressive, the point here is not to make things invalid just for the sake of it. Please engage charitably to foster productive discussions.

To your point, it's my understanding that `nLockTime` makes for a good extranonce because it's serialized at the very end of the coinbase transaction. Rolling it does not require re-hashing the whole transaction as you can just cache a midstate for the first 64 bytes. Could you provide more details about how you expect it to be used exactly, and under what circumstances the operation it improves may become a bottleneck?

Reply:
>@luke-jr why would miners switch to something different from the current extraNonce?

There is no "current extraNonce". Extranonces can be anywhere in theory. But this particular location, as @darosior says, minimises the need to re-hash a lot of data to roll it. That means it can more practically be done on ASICs or MCUs with nearly no latency.

Reply:
Can you do the same with an output? Since a coinbase doesn't have any inputs to sign, rolling an `OP_RETURN` output should be just as easy and there's no space limit?

Reply:
@luke-jr i see how it might be useful in theory but i don't think it would be in realistic conditions. Could you describe the specifics of a future ASIC design where it would matter?

Reply:
I'm moving this discussion [to the Delving thread](https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710/85?u=antoinep). Please @luke-jr provide more details there so your objection can be addressed.

Reply:
I don't use Delving

Reply:
Then can you please share details about your objection *somewhere*? Anywhere. Here, the dev mailing list, not Delving, 0bin, whatever.

Just sharing vague concerns without details makes it really hard to address your objection. And i trust it is not your intention to prevent having an objective discussion about the tradeoff being made here.